docs/design/resource-states.md

```diff
@@ -0,0 +1,221 @@
+  fields means it's easier for bugs to creep in where we set one but didn't clear another.
+* The `State` field is not widely used by all resources, and seems to conflict with `Provisioning`/`Provisioned`/`FailedProvisioning`. It's not clear when you should look at one or the other.
+
+### ASO v2 (autogenerated)
+The autogenerated resources don't currently have anything in their `Status` which conveys the Status _of the operator_. For the initial alpha releases, these fields are instead set as annotations on the object:
+```go
+	DeploymentIDAnnotation   = "deployment-id.infra.azure.com"
+	DeploymentNameAnnotation = "deployment-name.infra.azure.com"
+	ResourceStateAnnotation  = "resource-state.infra.azure.com"
+	ResourceErrorAnnotation  = "resource-error.infra.azure.com"
+```
+
+This was done mostly to avoid the additional work to code-generate the correct fields in each resources `status` and set them. Since it's possible that the whole `status` can get lost, some of the things may need to stay 
+as annotations (`deploymentId` possibly), while others are obviously status and can be derived again if lost. We took the temporary approach of putting it all in annotations while we learned more about what the best practices were.
+
+#### Problems with this apporach
+* The `ResourceStateAnnotation` is reusing the same enum to describe state as the `armclient` uses for deployments. This is problematic because that only supports `Succeeded`, `Failed`, `Deleting`, and `Accepted`.
+  `Succeeded`, `Failed`, and `Deleting` make sense in the context of the operator but `Accepted` doesn't really. There's no state for `InProgress` or `Reconciling`.
+* There's an awkward tension between sharing the state of the ARM deployment and the state of the resource. They're often the same but sometimes not (such as in the case the deployment is being deleted but the resource is not).
+* Using annotations is awkward because changing an annotation triggers another reconcile.
+* Using annotations means that it's not clear to the customer what the possible valid states are for the fields that are enums.
+* Users expect to look in `Status` for status related fields. Having it exposed as an annotation was never intended as a long term solution, just a hack so that we could get things up and running.
+
+## Propsal
+
+### Goals
+* The status of the operator should be reported as a subsection in the `Status` of the resource itself. See [#1504](https://github.com/Azure/azure-service-operator/pull/1504).
+* It should be clear to customers where to look to determine what the current status of their resource is.
+* The status should include information in both normal and failure cases.
+* When a failure occurs, it should be clear to customers that a failure has happened and what the cause of the failure was.
```

Comment:
```suggestion
* When a failure occurs, it should be clear to users that a failure has happened and what the cause of the failure was.
```

Reply:
would it make sense to also add as a goal `When a failure occurs, it should be clear to users whether that failure is transient and the operator can self-recover.` or something like that? to distinguish terminal (eg. VM size unavailable vs transient errors (eg. internal server error, will retry)

Reply:
yes! This definitely was a goal but wasn't explicitly called out - thanks.